crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+use std::io::{Cursor, Read};
+use std::sync::Arc;
+
+use itertools::Itertools;
+use zip::{read::ZipFile, result::ZipResult, ZipArchive};
+
+use crate::file_system::{FileRevision, FileSystem, FileSystemPath, FileType, Metadata, Result};
+
+#[derive(Debug, Clone)]
+pub struct VendoredFileSystem {
+    inner: VendoredZipArchive,
+}
+
+impl VendoredFileSystem {
+    const READ_ONLY_PERMISSIONS: u32 = 0o444;
+
+    pub fn new(raw_bytes: Arc<[u8]>) -> Result<Self> {
+        Ok(Self {
+            inner: VendoredZipArchive::new(raw_bytes)?,
+        })
+    }
+
+    /// Returns an owned copy of the underlying zip archive,
+    /// that can be queried for information regarding contained files and directories
+    fn vendored_archive(&self) -> VendoredZipArchive {
+        // Zip lookups require a mutable reference to the archive,
+        // which creates a nightmare with lifetimes unless you own the archive, hence the clone.
+        //
+        // The zip documentation states that:
+        //
+        //     At the moment, this type is cheap to clone if this is the case for the reader it uses.
+        //     However, this is not guaranteed by this crate and it may change in the future.
+        //
+        // So this should be fairly cheap *for now*.
+        self.inner.clone()
+    }
+
+    /// Create a new empty string, read the contents of `zipfile` into it, and return the string
+    fn read_zipfile(mut zipfile: ZipFile) -> Result<String> {
+        let mut buffer = String::with_capacity(zipfile.size() as usize);
+        zipfile.read_to_string(&mut buffer)?;
+        Ok(buffer)
+    }
+}
+
+impl FileSystem for VendoredFileSystem {
```

Comment:
We shouldn't implement `FileSystem` for vendored paths. I think that's also the part that @carljm felt some confusion around. There are a few reasons:

* `FileSystem` is an abstraction over `std::fs`. It's not a way of implementing an overlay filesystem that supports reading files from different sources. That's that `Vfs` does
* `FileSystem` works on `FileSystemPath`, but we want to use `VendoredFilePath` and `VendoredFilePathBuf` for vendored file. We intentionally use different types to prevent that someone tries to read a `VendoredPath` from the file system or the other way round
* Many `FileSystem` operations are simply irrelevant for `VendoredPath`s. The only thing that we need to support for vendored path is testing if a file exists (or a simplified metadata that returns the last modified date), and reading the content. We won't need file watching, directory traversal, or operations for testing if something's a directory. We only care about the files, and nothing else

That's why I think we can either implement the methods as free-standing functions similar to `fs::read_to_string`, or as a struct, but where we define our own functions.

the vendored path because vendored path should be represented using `VendoredPath` and `VendoredPathBuf`. 

The other reason is that most operations aren't releve